cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitor.java

```diff
@@ -82,12 +97,17 @@
   private final long pollTimeMillis;
   private final long gracefulShutdownMillis;
   private final List<MonitorMessage> lastProgramStateMessages;
+  private final DatasetFramework dsFramework;
+  private final Transactional transactional;
+  private final RetryStrategy retryStrategy;
   private volatile Thread runThread;
   private volatile boolean stopped;
+  private volatile boolean killed;
   private long programFinishTime;
 
   public RuntimeMonitor(ProgramRunId programRunId, CConfiguration cConf,
-                        MessagingService messagingService, ClientConfig clientConfig) {
+                        MessagingService messagingService, ClientConfig clientConfig,
+                        DatasetFramework datasetFramework, TransactionSystemClient txClient) {
     this.programRunId = programRunId;
     this.cConf = cConf;
     this.messagePublisher = new MultiThreadMessagingContext(messagingService).getMessagePublisher();
```

Comment:
Shouldn't get the publisher here if you want to do transactional publishing. You need to call `getMessagePublisher` inside the thread that executes the transaction. The easiest place is just to call it inside the `TxRunnable` / `TxCallable`.